### PR TITLE
App Language preference

### PR DESCRIPTION
**Pull request by choykarl**
#348

### CODE REVIEW

**Review by CanglongCl — DISMISSED** [review]

**Review by tisfeng — APPROVED** [review]
Looks good.

**Review by CanglongCl — DISMISSED** [review]

File: Easydict/NewApp/LanguagePreference/EZI18nHelper.swift [hunk 1]
```diff
@@ -0,0 +1,42 @@
...
+//  Created by choykarl on 2024/3/4.
+//  Copyright © 2024 izual. All rights reserved.
+//
+
+import SwiftUI
+
+// MARK: - EZI18nHelper
+
+@objcMembers
+class EZI18nHelper: NSObject {
```
Comment (phlpsong) [inline Easydict/NewApp/LanguagePreference/EZI18nHelper.swift:14]: No need to have the namespace prefix `EZ` in Swift.

File: Easydict/NewApp/LanguagePreference/EZI18nHelper.swift [hunk 2]
```diff
@@ -0,0 +1,42 @@
...
+
+    var localizedBundle: Bundle {
+        let res = localizeCode
+        let path = Bundle.main.path(forResource: res, ofType: "lproj")
+        let bundle: Bundle
+        if let path = path {
+            bundle = Bundle(path: path) ?? .main
+        } else {
+            bundle = .main
+        }
```
Comment (phlpsong) [inline Easydict/NewApp/LanguagePreference/EZI18nHelper.swift:26]:
```suggestion
        var bundle: Bundle = .main
        if let path {
            bundle = Bundle(path: path) ?? .main
        }
```

File: Easydict/NewApp/LanguagePreference/EZI18nHelper.swift [hunk 3]
```diff
@@ -0,0 +1,42 @@
...
+//  Copyright © 2024 izual. All rights reserved.
+//
+
+import SwiftUI
+
+// MARK: - EZI18nHelper
+
+@objcMembers
+class EZI18nHelper: NSObject {
+    static let languagePreferenceChangedNotification = "EZLanguagePreferenceChangedNotification"
```
Comment (phlpsong) [inline Easydict/NewApp/LanguagePreference/EZI18nHelper.swift:15]: It would be better to define this notification name in `Notification+Name.swift`.

**Review by phlpsong — CHANGES_REQUESTED** [review]

File: Easydict/NewApp/LanguagePreference/EZLocalizedBundle.swift [hunk 4]
```diff
@@ -0,0 +1,15 @@
...
+//  EZLocalizedBundle.swift
+//  Easydict
+//
+//  Created by choykarl on 2024/3/23.
+//  Copyright © 2024 izual. All rights reserved.
+//
+
+import Foundation
+
+class EZLocalizedBundle: Bundle {
```
Comment (phlpsong) [inline Easydict/NewApp/LanguagePreference/EZLocalizedBundle.swift:11]: Same as `I18nHelper`, should not have the namespace prefix `EZ` in Swift.

Comment (choykarl) [reply Easydict/NewApp/LanguagePreference/EZLocalizedBundle.swift]: you are right

File: Easydict/Feature/ViewController/Window/BaseQueryWindow/EZBaseQueryViewController.m [hunk 5]
```diff
@@ -192,6 +192,12 @@ - (void)setupUI {
             [self updateAllResultCellHeight];
         }];
     }];
+    
+    [defaultCenter addObserver:self selector:@selector(modifyLanduage:) name:EZI18nHelper.languagePreferenceChangedNotification object:nil];
```
Comment (phlpsong) [inline Easydict/Feature/ViewController/Window/BaseQueryWindow/EZBaseQueryViewController.m:196]:
```suggestion
    [defaultCenter addObserver:self selector:@selector(modifyLanduage:) name:NSNotification.languagePreferenceChanged object:nil];
```

File: Easydict/NewApp/LanguagePreference/LanguageState.swift [hunk 6]
```diff
@@ -0,0 +1,41 @@
...
+//
+//  Created by choykarl on 2024/3/3.
+//  Copyright © 2024 izual. All rights reserved.
+//
+
+import SwiftUI
+
+// MARK: - LanguageState
+
+let kEZLanguagePreferenceLocalKey = "kEZLanguagePreferenceLocalKey"
```
Comment (phlpsong) [inline Easydict/NewApp/LanguagePreference/LanguageState.swift:13]: I think this constant string looks like objc style definition, maybe move to objc side or rename to like `LanguagePreferenceLocalKey`.

File: Easydict/NewApp/LanguagePreference/I18nHelper.swift [hunk 7]
```diff
@@ -0,0 +1,39 @@
...
+//
+
+import SwiftUI
+
+// MARK: - I18nHelper
+
+@objcMembers
+@objc(EZI18nHelper)
+class I18nHelper: NSObject {
+    static let languagePreferenceChangedNotification = "EZLanguagePreferenceChangedNotification"
```
Comment (phlpsong) [inline Easydict/NewApp/LanguagePreference/I18nHelper.swift:16]:
```suggestion
    static let languagePreferenceChangedNotification = "LanguagePreferenceChangedNotification"
```

File: Easydict/NewApp/LanguagePreference/I18nHelper.swift [hunk 8]
```diff
@@ -0,0 +1,39 @@
+//
+//  EZI18nHelper.swift
```
Comment (phlpsong) [inline Easydict/NewApp/LanguagePreference/I18nHelper.swift:2]:
```suggestion
//  I18nHelper.swift
```

**Review by phlpsong — CHANGES_REQUESTED** [review]

**Review by phlpsong — APPROVED** [review]
Looks great!

**Comment by tisfeng** [comment]
Good review.

**Review by tisfeng — APPROVED** [review]

**Comment by tisfeng** [comment]
Great job!

**Pull request merged**